pkg/apiserver/etcd_config.go

```diff
@@ -115,7 +117,9 @@ func (c completedEtcdConfig) NewServer(stopCh <-chan struct{}) (*ClusterOperator
 		groupInfos = append(groupInfos, groupInfo)
 	}
 
-	caGroupInfo := ca_apis.GetClusterAPIBuilder().Build(roFactory)
+	caAPIBuilder := ca_apis.GetClusterAPIBuilder()
+	validation.ReplaceStorageBuilder(caAPIBuilder, ca.InternalMachine, validation.ReplaceMachineStorageBuilder)
```

Comment:
And we would repeat here for the API types we're more interested in which are Cluster and MachineSet and soon MachineDeployment.

These probably all have valid upstream deployment validation (for example: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/apis/cluster/v1alpha1/cluster_types.go#L112) which we'd want to combine with our validation on the providerConfig using: https://github.com/openshift/cluster-operator/blob/master/pkg/apis/clusteroperator/validation/cluster.go

Does this sound right?

Reply:
And add the check on MachineSet for a label linking it to a cluster.

Reply:
Yes.